tidy3d/components/boundary.py

```diff
@@ -0,0 +1,626 @@
+        """Error if a Bloch boundary is applied on only one side."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_bloch = isinstance(plus, BlochBoundary) + isinstance(minus, BlochBoundary)
+        if num_bloch == 1:
+            raise SetupError(
+                "Bloch boundaries must be applied either on both sides or on neither side."
+            )
+        return values
+
+    @pd.root_validator
+    def periodic_with_pml(cls, values):
+        """Error if PBC is specified with a PML."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_pbc = isinstance(plus, Periodic) + isinstance(minus, Periodic)
+        num_pml = isinstance(plus, (PML, StablePML, Absorber)) + isinstance(
+            minus, (PML, StablePML, Absorber)
+        )
+        if num_pbc == 1 and num_pml == 1:
+            raise SetupError("Cannot have both PML and PBC along the same dimension.")
+        return values
+
+    @pd.root_validator
+    def periodic_with_pec_pmc(cls, values):
+        """If a PBC is specified along with PEC or PMC on the other side, manually set the PBC
+        to PEC or PMC so that no special treatment of halos is required."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+
+        if isinstance(minus, (PECBoundary, PMCBoundary)) and isinstance(plus, Periodic):
+            plus = minus
+        elif isinstance(plus, (PECBoundary, PMCBoundary)) and isinstance(minus, Periodic):
+            minus = plus
+        values.update({"plus": plus, "minus": minus})
+        return values
+
+    @classmethod
+    def periodic(cls):
+        """Periodic boundary specification on both sides along a dimension.
+
+        Example
+        -------
+        >>> pbc = Boundary.periodic()
+        """
+        plus = Periodic()
+        minus = Periodic()
+        return cls(plus=plus, minus=minus)
+
+    @classmethod
+    def bloch(cls, bloch_vec: Complex):
+        """Bloch boundary specification on both sides along a dimension.
+
+        Parameters
+        ----------
+        bloch_vec : `Complex`
+            Normalized component of the Bloch vector in units of 2 * pi / (size along dimension)
+            in the background medium, along the dimension in which the boundary is specified.
+
+        Example
+        -------
+        >>> bloch = Boundary.bloch(bloch_vec=1)
+        """
+        plus = BlochBoundary(bloch_vec=bloch_vec)
+        minus = BlochBoundary(bloch_vec=bloch_vec)
+        return cls(plus=plus, minus=minus)
+
+    @classmethod
+    def bloch_from_source(
+        cls, source: BlochSourceType, domain_size: float, axis: Axis, medium: Medium = None
+    ):
+        """Bloch boundary specification on both sides along a dimension based on a given source.
+
+        Parameters
+        ----------
+        source : Union[:class:`GaussianBeam`, :class:`ModeSource`, :class:`PlaneWave`]
+            Angled source.
+        domain_size: float
```

Comment:
what about the center of the domain?  Where is that located?

Reply:
I'm not sure if the center is needed? Unless I'm missing something?

Reply:
ok, I got confused with PEC.  So this is the distance between the edges?  can this information be possibly taken from the Source or Simulation?  What happens if someone messes up and supplies a distance that is different from the actual simulation size?

Reply:
Yea it's the physical domain size along that dimension. If an incorrect size is provided, the Bloch BC won't work correctly. The source could have a size of `td.inf`, or a finite size smaller than the domain, so I'm not sure that'll work. The other option is to take `sim` as an argument, but that's a bit weird because then we'd have to define the simulation, then define the boundary, then redefine the simulation with the boundary spec. Maybe there's another way, e.g. take it from the grid spec?

Reply:
hm.  wonder if this is just something that should be computed server-side.  For example, just supply an axis.  Which size needs to be provided? the `sim.size[axis]` or the one after discretizing and snapping to grid? (if applicable)?

Reply:
In the new grid spec, the grid is snapped to the desired physical size of the domain even for a uniform mesh, so both approaches should be equivalent. Previously, this was done in the backend like you suggested, but after discussing with Momchil, he suggested that the fundamental variable of the Bloch BC should be the Bloch vector normalized in terms of the domain size. This corresponds to the bandstructure normalization provided by Lumerical and is maybe a more natural way of defining the Bloch vector for a photonic crystal, for example. This requires knowing the domain size beforehand though. Maybe we can reconsider / discuss further? @momchil-flex